### PR TITLE
[93] New Docker command to generate DockerFiles of specified services

### LINKED ISSUE #12: [schematics] uniform files structure
## Description

Presently the static files in the `logging` and `service-temporal` folder don't follow the same structure, in one the only the loggin folder is present while the other uses a `services/temporal` style of naming. The end goal is to have the cli generate the `logging` and `temporal` folders inside the root level `services` folder so the method used should be same and uniform for both services.

### Screenshots for reference
Screenshot 2024-01-03 at 11 27 53 PM

Screenshot 2024-01-03 at 11 28 14 PM

### PR DESCRIPTION
**Pull request by Savio629**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: https://github.com/SamagraX-Stencil/stencil/issues/135#event-13223410367

schematics link: https://github.com/SamagraX-Stencil/schematics/pull/10

## What is the new behavior?
`stencil docker <services...>`
where `<services>` = monitoring logging temporal 

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Comment by Savio629** [comment]
Updated existing test and implemented e2e test for docker command

**Comment by Savio629** [comment]
### Completed:
- Implemented Docker Command
    - Tooling specific setup
    - À la Carte setup
- Updated stencil cli docs - Docker command
- Updated existing test and implemented e2e test for docker command
- `--path` argument in À la Carte setup
- Updated docker-start.sh
- Integrate the `--docker` Flag

**Comment by Savio629** [comment]
README.md for Docker Command


## Docker Command

```
stencil docker <services...> [options]
stencil do <services...> [options]
```
**Description**

Creates a docker service/container for given command.
-  Tooling specific setup
    - Creates a folder with the given `<service>` inside docker directory
- À la Carte setup / Adhoc setup
    - Creates a docker compose or updates existing docker compose with desired `<service>`

**Arguments**

| Argument  |  Description |
|-----------|--------------|
|  `<service>`	 | The name of the new project |

**Options**

| Option  |  Description |
|-----------|--------------|
|  `--path [path]`	 | The path where the docker compose will be generated |


Note: Docker command supports multiple services at a time.

**Services**

-  Tooling specific setup


| Name  | Alias  |  Description |
|---|---|---|
|  `logging` | `lg`  | Generates a docker service for logging |
|  `monitoringService` |`ms`   | Generates a docker service for monitoring  |
|  `temporal` |  `tp` |  Generates a docker service for temporal |

- À la Carte setup / Adhoc setup


| Name  | Alias  |  Description |
|---|---|---|
|  `postgres` | `pg`  | Generate a docker compose for postgres |
|  `hasura` |`hs`   | Generate a docker compose for hasura  |

## How to include new docker services ?

**Stencil-cli**

1. Include the docker service in `lib/schematics/nest.collection.ts` with its name, alias and description.

**Schematics**
1. Create a folder inside of **schematics** package under `src/lib/docker/nameOfDockerService`
2. If the dockerService is a tooling setup then refer existing tooling setups such as temporal,monitoringService, logging.
     - Create `src/lib/docker/files/ts/nameOfDockerService` and paste all the necessary files needed to be generated when the service is called.
     - Create factory file, schema file and interface of the dockerService inside `src/lib/docker/nameOfDockerService` by refering existing tooling setup.
  3. If the dockerService is a adhoc setup then refer existing adhoc setup such as postgres, hasura.
     - Create factory file, schema file and interface of the dockerService inside `src/lib/docker/nameOfDockerService` by refering existing adhoc setup.
     - Paste the `docker-compose` and `.env` content inside of factory file refering existing adhoc setup.

## Which files will be changed/updated?

**Tooling setup**
- Basically whenever tooling setup is generated eg. `stencil docker temporal` , `docker/temporal` will be created.

**Adhoc Setup**
- Whenever adhoc setup is generated eg. `stencil docker postgres` , then` docker-compose`, `.env` and `docker-start.sh` is generated/updated.

File: Docker-readme.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+## Docker Command
+
+```
+stencil docker <services...> [options]
+stencil do <services...> [options]
+```
+**Description**
+
+Creates a docker service/container for given prompt.
```
Comment (techsavvyash) [inline Docker-readme.md]: are we supporting prompts?

File: actions/docker.action.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  private manager!: AbstractPackageManager;
+  public async handle(commandInputs: Input[],options: Input[]) {
+    this.manager = await PackageManagerFactory.find();
+    
+    const collection: AbstractCollection = CollectionFactory.create(
+       Collection.NESTJS,
+    );
+    const schematicOptions: SchematicOption[] = this.mapSchematicOptions(commandInputs);
+    schematicOptions.push(
+      new SchematicOption('language', 'ts'),
```
Comment (techsavvyash) [inline actions/docker.action.ts:31]: why have we hardcoded to `ts`

File: actions/new.action.ts [hunk 3]
```diff
@@ -61,21 +64,17 @@ export class NewAction extends AbstractAction {
...
     const shouldInitializeTemporal = options.some(
       (option) => option.name === 'temporal' && option.value === 'yes',
     );
 
-    const shouldInitializeLogging = options.some(
-      (option) => option.name === 'logging' && option.value === 'yes',
-    );
+    // const shouldInitializeLogging = options.some(
+    //   (option) => option.name === 'logging' && option.value === 'yes',
+    // );
```
Comment (techsavvyash) [inline actions/new.action.ts]: remove comments

File: actions/new.action.ts [hunk 4]
```diff
@@ -110,30 +109,25 @@ export class NewAction extends AbstractAction {
...
-    await createLogging(
-      isDryRunEnabled as boolean,
-      projectDirectory,
-      shouldInitializeLogging as boolean,
-    );
+    // await createLogging(
+    //   isDryRunEnabled as boolean,
+    //   projectDirectory,
+    //   shouldInitializeLogging as boolean,
+    // );
```
Comment (techsavvyash) [inline actions/new.action.ts]: remove comments

File: actions/new.action.ts [hunk 5]
```diff
@@ -145,7 +139,7 @@ export class NewAction extends AbstractAction {
       if (!shouldSkipGit) {
         await initializeGitRepository(projectDirectory);
         await createGitIgnoreFile(projectDirectory);
-        await createRegistry(projectDirectory, shouldInitializePrima,shouldInitializeUserService,shouldInstallMonitoring, shouldInitializeMonitoring,shouldInitializeTemporal,shouldInitializeLogging,shouldInitializeFileUpload);
+        await createRegistry(projectDirectory, shouldInitializePrima,shouldInitializeUserService,shouldInitializeMonitoring,shouldInitializeTemporal,shouldInitializeFileUpload);
```
Comment (techsavvyash) [inline actions/new.action.ts:135]: why are we removing logging everywhere?

File: actions/new.action.ts [hunk 6]
```diff
@@ -176,14 +170,14 @@ const getFixturesInput = (inputs: Input[]) =>
 const getMonitoringInput = (inputs: Input[]) =>
   inputs.find((options) => options.name === 'monitoring');
 
-const getMonitoringServiceInput = (inputs: Input[]) =>
-  inputs.find((options) => options.name === 'monitoringService');
+// const getMonitoringServiceInput = (inputs: Input[]) =>
+//   inputs.find((options) => options.name === 'monitoringService');
```
Comment (techsavvyash) [inline actions/new.action.ts]: remove comments

File: actions/new.action.ts [hunk 7]
```diff
@@ -176,14 +170,14 @@ const getFixturesInput = (inputs: Input[]) =>
...
 
 const getTemporalInput = (inputs: Input[]) =>
   inputs.find((options) => options.name === 'temporal');
 
-const getLoggingInput = (inputs: Input[]) =>
-  inputs.find((options) => options.name === 'logging');
+// const getLoggingInput = (inputs: Input[]) =>
+//   inputs.find((options) => options.name === 'logging');
```
Comment (techsavvyash) [inline actions/new.action.ts]: remove comments

File: actions/new.action.ts [hunk 8]
```diff
@@ -238,23 +232,23 @@ const askForMissingInformation = async (inputs: Input[], options: Input[]) => {
...
-  const monitoringServiceInput = getMonitoringServiceInput(options);
-  if (!monitoringServiceInput!.value) {
-    const answers = await askForMonitoringService();
-    replaceInputMissingInformation(options, answers);
-  }
+  // const monitoringServiceInput = getMonitoringServiceInput(options);
+  // if (!monitoringServiceInput!.value) {
+  //   const answers = await askForMonitoringService();
+  //   replaceInputMissingInformation(options, answers);
+  // }
```
Comment (techsavvyash) [inline actions/new.action.ts]: remove comments

File: actions/new.action.ts [hunk 9]
```diff
@@ -238,23 +232,23 @@ const askForMissingInformation = async (inputs: Input[], options: Input[]) => {
...
-  const loggingInput = getLoggingInput(options);
-  if (!loggingInput!.value) {
-    const answers = await askForLogging();
-    replaceInputMissingInformation(options, answers);
-  }
+  // const loggingInput = getLoggingInput(options);
+  // if (!loggingInput!.value) {
+  //   const answers = await askForLogging();
+  //   replaceInputMissingInformation(options, answers);
+  // }
```
Comment (techsavvyash) [inline actions/new.action.ts]: remove comments

File: actions/new.action.ts [hunk 10]
```diff
@@ -488,35 +459,35 @@ const createTemporal = async (
...
+//     return;
+//   }
+
+//   const LoggingInstance = new ClassLogging();
+//   try {
+//     await LoggingInstance.create(createDirectory);
+//   } catch (error) {
+//     console.error('could not create the logging folder');
+//   }
+// };
```
Comment (techsavvyash) [inline actions/new.action.ts]: remove comments

**Review by techsavvyash — CHANGES_REQUESTED** [review]

File: Docker-readme.md (hunk 1 above)
Comment (Savio629) [reply Docker-readme.md]: Not yet

Comment (Savio629) [reply Docker-readme.md]: Will update the readme

File: actions/docker.action.ts (hunk 2 above)
Comment (Savio629) [reply actions/docker.action.ts:31]: this was already present with new command
It is the default language which is pushed if no language is given by the user

File: actions/new.action.ts (hunk 3 above)
Comment (Savio629) [reply actions/new.action.ts]: Wanted to confirm about removal of logging 
Will remove the comments

File: actions/new.action.ts (hunk 5 above)
Comment (Savio629) [reply actions/new.action.ts:135]: As of right now logging only make changes on docker level not on code level so I shifted logging to docker service command and removed from existing setup

Comment (Savio629) [reply actions/new.action.ts:135]: @techsavvyash

**Comment by Savio629** [comment]
- [x] Resolved all the comments
- [x] Added `minio` and `fusionauth` in the docker command
- [x] Tested the commands well

**Comment by Savio629** [comment]
image

**Review by techsavvyash — APPROVED** [review]
